src/lib/tecs/TECS.cpp

```diff
@@ -30,627 +30,682 @@
  * POSSIBILITY OF SUCH DAMAGE.
  *
  ****************************************************************************/
+/**
+ * @file TECS.cpp
+ *
+ * @author Paul Riseborough
+ */
 
 #include "TECS.hpp"
 
 #include <lib/geo/geo.h>
 
 #include <px4_platform_common/defines.h>
 
+#include "matrix/Matrix.hpp"
+#include "matrix/Vector2.hpp"
+
 using math::constrain;
 using math::max;
 using math::min;
 
-static constexpr float DT_MIN = 0.001f;	///< minimum allowed value of _dt (sec)
-static constexpr float DT_MAX = 1.0f;	///< max value of _dt allowed before a filter state reset is performed (sec)
+static inline constexpr bool TIMESTAMP_VALID(float dt) { return (PX4_ISFINITE(dt) && dt > FLT_EPSILON);}
 
-/**
- * @file TECS.cpp
- *
- * @author Paul Riseborough
- */
-
-/*
- * This function implements a complementary filter to estimate the climb rate when
- * inertial nav data is not available. It also calculates a true airspeed derivative
- * which is used by the airspeed complimentary filter.
- */
-void TECS::update_vehicle_state_estimates(float equivalent_airspeed, const float speed_deriv_forward,
-		bool altitude_lock, float altitude, float vz)
+void TECSAirspeedFilter::initialize(const float equivalent_airspeed)
 {
-	// calculate the time lapsed since the last update
-	uint64_t now = hrt_absolute_time();
-	float dt = fmaxf((now - _state_update_timestamp) * 1e-6f, DT_MIN);
 
-	bool reset_altitude = false;
 
-	if (_state_update_timestamp == 0 || dt > DT_MAX) {
-		dt = DT_DEFAULT;
-		reset_altitude = true;
-	}
-
-	if (!altitude_lock) {
-		reset_altitude = true;
-	}
+	_airspeed_state.speed = equivalent_airspeed;
+	_airspeed_state.speed_rate = 0.0f;
+}
 
-	if (reset_altitude) {
-		_states_initialized = false;
+void TECSAirspeedFilter::update(const float dt, const Input &input, const Param &param,
+				const bool airspeed_sensor_available)
+{
+	// Input checking
+	if (!TIMESTAMP_VALID(dt)) {
+		// Do not update the states.
+		PX4_WARN("Time intervall is not valid.");
+		return;
 	}
 
-	_state_update_timestamp = now;
-	_EAS = equivalent_airspeed;
+	float airspeed;
 
-	// Set the velocity and position state to the the INS data
-	_vert_vel_state = -vz;
-	_vert_pos_state = altitude;
-
-	// Update and average speed rate of change if airspeed is being measured
-	if (PX4_ISFINITE(equivalent_airspeed) && airspeed_sensor_enabled()) {
-		_tas_rate_raw = speed_deriv_forward;
-		// Apply some noise filtering
-		_TAS_rate_filter.update(speed_deriv_forward);
-		_tas_rate_filtered = _TAS_rate_filter.getState();
+	if (PX4_ISFINITE(input.equivalent_airspeed) && airspeed_sensor_available) {
+		airspeed = input.equivalent_airspeed;
 
 	} else {
-		_tas_rate_raw = 0.0f;
-		_tas_rate_filtered = 0.0f;
+		airspeed = param.equivalent_airspeed_trim;
```

Comment:
@KonradRudin When we don't have an airspeed sensor, can't we just set airspeed to trim, rate to 0 and then return and not run the filter?

Reply:
I did run the estimator in this case to avoid a bump, when flying with an airspeed sensor which could be disabled in time due to error, and thereby generating a smoother signal. But this is a corner case, so happy to discuss.